app/Platform/Controllers/CompareUnlocksController.php

```diff
@@ -0,0 +1,120 @@
+            abort(403);
+        }
+
+        $validatedData = $request->validate([
+            'sort' => 'sometimes|string|in:selfUnlocks,otherUnlocks,display,title',
+        ]);
+        $sortOrder = $validatedData['sort'] ?? 'display';
+
+        $achievements = [];
+        foreach ($game->achievements()->published()->get() as $achievement) {
+            $achievements[$achievement->ID] = $achievement->toArray();
+        }
+
+        $activeUserUnlocks = $activeUser->playerAchievements()
+            ->join('Achievements', 'Achievements.ID', '=', 'player_achievements.achievement_id')
+            ->where('Achievements.GameID', $game->ID)
+            ->select(['achievement_id', 'unlocked_at', 'unlocked_hardcore_at']);
+        foreach ($activeUserUnlocks->get() as $unlock) {
+            if (!array_key_exists($unlock->achievement_id, $achievements)) {
+                continue;
+            }
+
+            if ($unlock->unlocked_hardcore_at) {
+                $achievements[$unlock->achievement_id]['userTimestampRaw'] = $unlock->unlocked_hardcore_at;
+                $achievements[$unlock->achievement_id]['userTimestamp'] = Carbon::parse($unlock->unlocked_hardcore_at)->format('d M Y, g:ia');
+                $achievements[$unlock->achievement_id]['userHardcore'] = true;
+            } elseif ($unlock->unlocked_at) {
+                $achievements[$unlock->achievement_id]['userTimestampRaw'] = $unlock->unlocked_at;
+                $achievements[$unlock->achievement_id]['userTimestamp'] = Carbon::parse($unlock->unlocked_at)->format('d M Y, g:ia');
+                $achievements[$unlock->achievement_id]['userHardcore'] = false;
+            }
+        }
+
+        $otherUserUnlocks = $user->playerAchievements()
+            ->join('Achievements', 'Achievements.ID', '=', 'player_achievements.achievement_id')
+            ->where('Achievements.GameID', $game->ID)
+            ->select(['achievement_id', 'unlocked_at', 'unlocked_hardcore_at']);
+        foreach ($otherUserUnlocks->get() as $unlock) {
+            if (!array_key_exists($unlock->achievement_id, $achievements)) {
+                continue;
+            }
+
+            if ($unlock->unlocked_hardcore_at) {
+                $achievements[$unlock->achievement_id]['otherUserTimestampRaw'] = $unlock->unlocked_hardcore_at;
+                $achievements[$unlock->achievement_id]['otherUserTimestamp'] = Carbon::parse($unlock->unlocked_hardcore_at)->format('d M Y, g:ia');
+                $achievements[$unlock->achievement_id]['otherUserHardcore'] = true;
+            } elseif ($unlock->unlocked_at) {
+                $achievements[$unlock->achievement_id]['otherUserTimestampRaw'] = $unlock->unlocked_at;
+                $achievements[$unlock->achievement_id]['otherUserTimestamp'] = Carbon::parse($unlock->unlocked_at)->format('d M Y, g:ia');
+                $achievements[$unlock->achievement_id]['otherUserHardcore'] = false;
+            }
+        }
+
```

Comment:
I can't help but notice the code for managing `$activeUserUnlocks` and `$otherUserUnlocks` is largely duplicative.

I think we may be able to get away with something like:

```php
processUserUnlocks($activeUser, $gameId, $achievements, 'user');
processUserUnlocks($otherUser, $gameId, $achievements, 'otherUser');
```

```php
// pseudocode
function processUserUnlocks($user, $gameId, &$achievements, $prefix)
{
  $unlocks = ... make the Eloquent ORM query ...

  foreach ($unlocks as $unlock) {
    if (!array_key_exists(...) { continue; }

    $timestampKey = "{$prefix}TimestampRaw";
    $formattedTimestampKey = "{$prefix}Timestamp";
    $hardcoreKey = "{$prefix}Hardcore";

    if ($unlock->unlocked_hardcore_at) {
      $achievements[$unlock->achievement_id][$timestampKey] = $unlock->unlocked_hardcore_at;
      // ... more stuff ...
    } elseif ($unlock->unlocked_at) {
      // ...
    }
  }
}
```